### DISCUSSION TITLE
BTree Layout

### DISCUSSION THREAD

**Discussion by pjhfggij** [anametologin/krohnkite#33] [Ideas]
Thank you thousandfold for reviving krohnkite - this is a lifesaver! The past agonizing year since kde6 API break with all the unfortunate bugs in polonium made me appreciate one thing - the [BTree layout](https://zeroxoneafour.github.io/polonium/usage.html#BTree). It works surprisingly well and allows you to fit many more windows on a single screen than the traditional Tile (master/stack) layout. 

Would it be feasible to add a similar layout engine to krohnkite?

**Comment by anametologin** [comment]
Why not, binary tree layout looks good. I'll see how to do it well.

**Reply by pjhfggij** [reply]
Thank you thousandfold and one :)

**Comment by anametologin** [comment]
Added.

**Reply by pjhfggij** [reply]
That is fantastic, thank you! Nothing short of amazing how fast that was!!!
I've initially had some issues with activating it, but after toggling off all other layouts and somehow crashing Kwin in the process of enabling/disabling Korhnkite BTree Layout came out in all its glory. Perhaps this was caused by installing krohnkite first from KDE's menu and now via git clone/make install.
I can't resize the windows yet (shrink/grown height/width don't work).
One behavior that was useful in polonium was enlarging some widows by "moving other windows out". For example here:
![image](https://github.com/anametologin/krohnkite/assets/161205138/9739a171-5553-4bf5-be00-a678c607d92a)
moving window 3 to the right would make window 1 half the screen. However, I only used it to make a certain window bigger - I like the current way krohnkite handles moving windows already, so perhaps a better solution would be to "make master"work so that it "enlarges window" horizontally/vertically?

**Reply by anametologin** [reply]
> crashing Kwin

X11 or Wayland are you have?

> to make a certain window bigger

The ways to make the window bigger, here is what I managed to quickly come up with:
1.monocle
2. make it float
3. Kwin Maximize window
4. Kwin Maximize window Vertically
5. Kwin Maximize window Horizontally
6. make it master and switch to tile layout
7. making lower window 3 -  first window will half of the screen

**Reply by pjhfggij** [reply]
> X11 or Wayland

Wayland
neofetch:
```
                     ./o.                  user@host
                   ./sssso-                -----------
                 `:osssssss+-              OS: EndeavourOS Linux x86_64
               `:+sssssssssso/.            Host: 81YT Lenovo Legion 7 15IMH05
             `-/ossssssssssssso/.          Kernel: 6.9.5-arch1-1
           `-/+sssssssssssssssso+:`        Uptime: 19 hours, 49 mins
         `-:/+sssssssssssssssssso+/.       Packages: 2960 (pacman), 21 (flatpak)
       `.://osssssssssssssssssssso++-      Shell: zsh 5.9
      .://+ssssssssssssssssssssssso++:     Resolution: 1920x1080, 1920x1080
    .:///ossssssssssssssssssssssssso++:    DE: KDE
  `:////ssssssssssssssssssssssssssso+++.   WM: KWin
`-////+ssssssssssssssssssssssssssso++++-   WM Theme: One-Dark-Blue
 `..-+oosssssssssssssssssssssssso+++++/`   Theme: Breeze Dark [KDE], Infinity-GTK [GTK2/3]
   ./++++++++++++++++++++++++++++++/:.     Icons: Tela-blue-dark [KDE], Tela-blue-dark [GTK2/3]
  `:::::::::::::::::::::::::------``       Terminal: tmux
                                           CPU: Intel i7-10750H (12) @ 5.000GHz
                                           GPU: NVIDIA GeForce RTX 2070 SUPER Mobile / Max-Q
                                           GPU: Intel CometLake-H GT2 [UHD Graphics]
                                           Memory: 9919MiB / 31801MiB

````
> The ways to make the window bigger

I should have mentioned that for now I'm toggling the tile layout whenever a window needs to be bigger/become "master". It's effectively the same, but sometimes you don't want to extend the window only to a third of the screen or other arbitrary sizes.

**Comment by anametologin** [comment]
> sometimes you don't want to extend the window only to a third of the screen or other arbitrary sizes

I think I understand what you mean, it makes sense. I did binary tree layer as dashboard, because I need the dashboard, if someone realize "adjust" function in this layer that is will be cool.

**Reply by pjhfggij** [reply]
Yes, exactly, also sorry for writing it in a bit confusing way - I only meant that extending/shrinking windows in BTree layout would be nice. 
Another thing I think would be useful is to spawn new window next to the currently active one instead of always putting it in the bottom right corner.

**Comment by anametologin** [comment]
> I only meant that extending/shrinking windows in BTree layout would be nice.

try tree column layout

> always putting it in the bottom right corner

I think about add 3 ways in config:
1. adding to the end 
2. adding to second
3.  replace master

**Reply by pjhfggij** [reply]
> try tree column layout

Tried it, but there was something about it that led me to tile layout again (that I can't quite recall right now)

> I think about add 3 ways in config:

yep, they all sound useful to me
